test/functional/legacy/061_undo_tree_spec.lua

```diff
@@ -0,0 +1,186 @@
+-- Tests for undo tree and :eralier and :later.
+
+local helpers = require('test.functional.helpers')
+local feed, insert, source, eq, eval, clear, execute, expect, wait =
+  helpers.feed, helpers.insert, helpers.source, helpers.eq, helpers.eval,
+  helpers.clear, helpers.execute, helpers.expect, helpers.wait
+
+describe('the undo tree', function()
+  setup(clear)
+  teardown(function()
+    os.remove('Xtest')
+  end)
+
+  it('is working', function()
+    -- Assert that no undo history is present.
+    eq({}, eval('undotree().entries'))
+    insert([[
+      
+      123456789]])
+
+    -- Clear the undo history after the insertion (see :h clear-undo)
+    execute('let old_undolevels = &undolevels')
+    execute('set undolevels=-1')
+    feed('a <BS><Esc>') 
+    execute('let &undolevels = old_undolevels')
+    execute('unlet old_undolevels')
+    eq({}, eval('undotree().entries'))
+
+    -- Delete three characters and undo.
+    feed('Gxxx')
+    eq('456789', eval('getline(".")'))
+    feed('g-')
+    eq('3456789', eval('getline(".")'))
+    feed('g-')
+    eq('23456789', eval('getline(".")'))
+    feed('g-')
+    eq('123456789', eval('getline(".")'))
+    feed('g-')
+    eq('123456789', eval('getline(".")'))
+
+    -- Delete three other characters and go back in time step by step.
+    feed('$xxx')
+    eq('123456', eval('getline(".")'))
+    execute('sleep 1')
+    wait()
+    feed('g-')
+    eq('1234567', eval('getline(".")'))
+    feed('g-')
+    eq('12345678', eval('getline(".")'))
+    feed('g-')
+    eq('456789', eval('getline(".")'))
+    feed('g-')
+    eq('3456789', eval('getline(".")'))
+    feed('g-')
+    eq('23456789', eval('getline(".")'))
+    feed('g-')
+    eq('123456789', eval('getline(".")'))
+    feed('g-')
+    eq('123456789', eval('getline(".")'))
+    feed('g-')
+    eq('123456789', eval('getline(".")'))
+    feed('10g+')
+    eq('123456', eval('getline(".")'))
+
+    -- Delay for three seconds and go some seconds forward and backward.
+    execute('sleep 2')
+    wait()
+    feed('Aa<esc>')
+    feed('Ab<esc>')
+    feed('Ac<esc>')
+    eq('123456abc', eval('getline(".")'))
+    execute('ear 1s')
+    eq('123456', eval('getline(".")'))
+    execute('ear 3s')
+    eq('123456789', eval('getline(".")'))
+    execute('later 1s')
+    eq('123456', eval('getline(".")'))
+    execute('later 1h')
+    eq('123456abc', eval('getline(".")'))
+
+    -- Test undojoin.
+    feed('Goaaaa<esc>')
+    feed('obbbb<esc>u')
+    eq('aaaa', eval('getline(".")'))
+    feed('obbbb<esc>')
+    execute('undojoin')
+    feed('occcc<esc>u')
+    -- TODO At this point the original test will write "aaaa" to test.out.
+    -- Why is the line "bbbb" here?
+    eq('bbbb', eval('getline(".")'))
```

Comment:
The behavior in vim and nvim if I perform this step manually is the same as in this new test file. The two changes (`obbbb<esc>` and `occcc<esc>`) are not joined and can be undone separately.